File: hard/3534-path-existence-queries-in-a-graph-ii/solution.java
import java.util.Arrays;

/**
 * 3534. Path Existence Queries in a Graph II
 * Time: O((n + q) log n)
 * Space: O(n log n)
 */
class Solution {
    public int[] pathExistenceQueries(int n, int[] nums, int maxDiff, int[][] queries) {
        // Sort nodes by value; rank[node] = its position in sorted order.
        Integer[] order = new Integer[n];
        for (int i = 0; i < n; i++) {
            order[i] = i;
        }
        Arrays.sort(order, (a, b) -> nums[a] - nums[b]);
        int[] vals = new int[n];
        int[] rank = new int[n];
        for (int k = 0; k < n; k++) {
            vals[k] = nums[order[k]];
            rank[order[k]] = k;
        }

        // R[i] = farthest sorted index reachable in one hop from i.
        int[] R = new int[n];
        int j = 0;
        for (int i = 0; i < n; i++) {
            if (j < i) {
                j = i;
            }
            while (j + 1 < n && vals[j + 1] - vals[i] <= maxDiff) {
                j++;
            }
            R[i] = j;
        }

        // Component ids via consecutive gaps.
        int[] comp = new int[n];
        for (int i = 1; i < n; i++) {
            comp[i] = comp[i - 1] + (vals[i] - vals[i - 1] > maxDiff ? 1 : 0);
        }

        // Binary lifting on the farthest-jump function.
        int LOG = Math.max(1, 32 - Integer.numberOfLeadingZeros(Math.max(1, n)));
        int[][] up = new int[LOG][n];
        up[0] = R;
        for (int k = 1; k < LOG; k++) {
            for (int i = 0; i < n; i++) {
                up[k][i] = up[k - 1][up[k - 1][i]];
            }
        }

        int[] ans = new int[queries.length];
        for (int q = 0; q < queries.length; q++) {
            int su = rank[queries[q][0]], sv = rank[queries[q][1]];
            if (su > sv) {
                int t = su;
                su = sv;
                sv = t;
            }
            if (su == sv) {
                ans[q] = 0;
            } else if (comp[su] != comp[sv]) {
                ans[q] = -1;
            } else {
                int cur = su, steps = 0;
                for (int k = LOG - 1; k >= 0; k--) {
                    if (up[k][cur] < sv) {
                        cur = up[k][cur];
                        steps += 1 << k;
                    }
                }
                ans[q] = steps + 1; // one final jump reaches sv
            }
        }
        return ans;
    }
}

File: hard/3534-path-existence-queries-in-a-graph-ii/solution.py
from typing import List


class Solution:
    def pathExistenceQueries(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[int]:
        """
        3534. Path Existence Queries in a Graph II
        Time: O((n + q) log n)
        Space: O(n log n)
        """
        # Sort nodes by value; rank[node] = its position in sorted order.
        order = sorted(range(n), key=lambda i: nums[i])
        vals = [nums[i] for i in order]
        rank = [0] * n
        for k, node in enumerate(order):
            rank[node] = k

        # R[i] = farthest sorted index reachable in one hop from i.
        R = [0] * n
        j = 0
        for i in range(n):
            if j < i:
                j = i
            while j + 1 < n and vals[j + 1] - vals[i] <= maxDiff:
                j += 1
            R[i] = j

        # Component ids via consecutive gaps.
        comp = [0] * n
        for i in range(1, n):
            comp[i] = comp[i - 1] + (1 if vals[i] - vals[i - 1] > maxDiff else 0)

        # Binary lifting on the farthest-jump function.
        LOG = max(1, n.bit_length())
        up = [R]  # up[0]
        for _ in range(1, LOG):
            prev = up[-1]
            cur = [prev[prev[i]] for i in range(n)]
            up.append(cur)

        ans = []
        for u, v in queries:
            su, sv = rank[u], rank[v]
            if su > sv:
                su, sv = sv, su
            if su == sv:
                ans.append(0)
            elif comp[su] != comp[sv]:
                ans.append(-1)
            else:
                cur = su
                steps = 0
                for k in range(LOG - 1, -1, -1):
                    if up[k][cur] < sv:
                        cur = up[k][cur]
                        steps += 1 << k
                ans.append(steps + 1)  # one final jump reaches sv
        return ans
